operator/controllers/execution/scheduledscan_controller.go

```diff
@@ -156,12 +156,46 @@ func (r *ScheduledScanReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		}
 
 		// Recalculate next schedule
-		nextSchedule = time.Now().Add(scheduledScan.Spec.Interval.Duration)
+		nextSchedule, err = getNextSchedule(r, scheduledScan, time.Now())
 	}
 
 	return ctrl.Result{RequeueAfter: nextSchedule.Sub(time.Now())}, nil
 }
 
+func getNextSchedule(r *ScheduledScanReconciler, scheduledScan executionv1.ScheduledScan, now time.Time) (next time.Time, err error) {
+	// check if the Cron schedule is set
+	if scheduledScan.Spec.Schedule != "" {
+		sched, err := cron.ParseStandard(scheduledScan.Spec.Schedule)
+		if err != nil {
+			return time.Time{}, fmt.Errorf("Unparseable schedule %q: %v", scheduledScan.Spec.Schedule, err)
+		}
+
+		// for optimization purposes, cheat a bit and start from our last observed run time
+		// we could reconstitute this here, but there's not much point, since we've
+		// just updated it.
+		var earliestTime time.Time
+		if scheduledScan.Status.LastScheduleTime != nil {
+			earliestTime = scheduledScan.Status.LastScheduleTime.Time
+		} else {
+			earliestTime = scheduledScan.ObjectMeta.CreationTimestamp.Time
+		}
+		if earliestTime.After(now) {
+			return sched.Next(now), nil
+		}
+		return sched.Next(earliestTime), nil
+	}
+	if scheduledScan.Spec.Interval.Duration > 0 {
+		var nextSchedule time.Time
+		if scheduledScan.Status.LastScheduleTime != nil {
+			nextSchedule = scheduledScan.Status.LastScheduleTime.Add(scheduledScan.Spec.Interval.Duration)
+		} else {
+			nextSchedule = time.Now().Add(-1 * time.Second)
+		}
+		return nextSchedule, nil
+	}
+	return time.Time{}, fmt.Errorf("No schedule or interval found")
+}
```

Comment:
Do we have a (easy) way to validate that the user set either the cron or the interval?
Or at least add a error event to the scheduled scan to indicate to users that the config is invalid without them having to look through the operator logs?

Reply:
I can turn this into 
```
if (Schedule ) 
...
else if (Interval)
... 
else 
return time.Time{}, fmt.Errorf("No schedule or interval found")
```
Would that be good enough ?

Reply:
Mhhh isn't that basically what we have now?
The problem is that for users without access to the logs of the operator these kinds of errors would be completly invisible. Ideally these invalid ScheduledScans could not even get created. If that's to hard (because it'd probably require a custom validating webhook we could at least add attach a error event to the ScheduledScan so that users using `kubectl describe scheduledscan <name>` could at least see a hint whats wrong with it.

Reply:
We do that for Scans where the scan type doesn't exist (as far as i remember)

Reply:
Just pushed a possible implementation. I am running however into a problem. The events are not getting recorded due to permission issues. I get the following error log:
`
│ E0714 10:01:20.230395       1 event.go:267] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"old-typo3.demo-targets.svc.org-1min.1771b311efbe8e98", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ZZZ_DeprecatedClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"ScheduledScan", Namespace:"default", Name:"old-typo3.demo-targets.svc.org-1min", UID:"e17ce9b3-4b7c-4d92-8157-3cea39efd8cb", APIVersion:"execution.securecodebox.io/v1", ResourceVersion:"1642", FieldPath:""}, Reason:"ScheduleParseError", Message:"Unparseable schedule \"/5 * * * *\": Failed to parse int from : strconv.Atoi: parsing \"\": invalid syntax", Source:v1.EventSource{Component:"ScheduledScanController", Host:""}, FirstTimestamp:time.Date(2023, time.July, 14, 9, 50, 24, 858705560, time.Local), LastTimestamp:time.Date(2023, time.July, 14, 10, 1, 20, 228309484, time.Local), Count:18, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events "old-typo3.demo-targets.svc.org-1min.1771b311efbe8e98" is forbidden: User "system:serviceaccount:securecodebox-system:securecodebox-operator" cannot patch resource "events" in API group "" in the namespace "default"' (will not retry!) 
`


TLDR:
`
User "system:serviceaccount:securecodebox-system:securecodebox-operator" cannot patch resource "events" in API group "" in the namespace "default"' (will not retry!)                                 
`

I am wondering why are the permissions missing here (or not functioning correctly) but supposedly work here:
https://github.com/secureCodeBox/secureCodeBox/blob/main/operator/controllers/execution/scantype_controller.go#L87
Any idea?

Reply:
Reason: https://github.com/secureCodeBox/secureCodeBox/pull/1811